File: Source/System/Localization.c
// LOCALIZATION.C
// (C) 2025 Iliyas Jorio
// This file is part of Nanosaur 2. https://github.com/jorio/nanosaur2

// LoadLocalizedStrings/Localize/DisposeLocalizedStrings are now in
// Localization.swift. LocalizeWithPlaceholder stays here because it takes
// C variadic arguments, which Swift can't implement. IsNativeEnglishSystem
// and GetBestLanguageIDFromSystemLocale also stay here since they consume
// kLanguageCodesISO639_1, a 2D fixed char array.

#include "game.h"

static const char kLanguageCodesISO639_1[NUM_LANGUAGES][3] =
{
	[LANGUAGE_ENGLISH	] = "en",
	[LANGUAGE_FRENCH	] = "fr",
	[LANGUAGE_GERMAN	] = "de",
	[LANGUAGE_SPANISH	] = "es",
	[LANGUAGE_ITALIAN	] = "it",
	[LANGUAGE_SWEDISH	] = "sv",
	[LANGUAGE_DUTCH		] = "nl",
	[LANGUAGE_RUSSIAN	] = "ru",
};

int LocalizeWithPlaceholder(LocStrID stringID, char* buf0, size_t bufSize, const char* format, ...)
{
	char* buf = buf0;
	const char* localizedString = Localize(stringID);

	const char* placeholder = SDL_strchr(localizedString, '#');

	if (!placeholder)
	{
		goto fail;
	}

	size_t bytesBeforePlaceholder = placeholder - localizedString;

	if (bytesBeforePlaceholder + 1 >= bufSize)		// +1 for nul terminator
	{
		goto fail;
	}

	SDL_memcpy(buf, localizedString, bytesBeforePlaceholder);
	buf += bytesBeforePlaceholder;
	bufSize -= bytesBeforePlaceholder;

	va_list args;
	va_start(args, format);
	int rc = SDL_vsnprintf(buf, bufSize, format, args);
	va_end(args);

	if (rc >= 0)
	{
		buf += rc;
		bufSize -= rc;

		rc = SDL_snprintf(buf, bufSize, "%s", placeholder + 1);
		if (rc >= 0)
		{
			buf += rc;
			bufSize -= rc;
		}
	}

	return (int) (buf - buf0);

fail:
	return SDL_snprintf(buf, bufSize, "%s", localizedString);
}

bool IsNativeEnglishSystem(void)
{
	bool prefersEnglish = true;

	int numLocales = 0;
	SDL_Locale** localeList = SDL_GetPreferredLocales(&numLocales);

	if (NULL != localeList
		&& 0 != localeList[0]->language
		&& (0 != strncmp(localeList[0]->language, kLanguageCodesISO639_1[LANGUAGE_ENGLISH], 2)))
	{
		prefersEnglish = false;
	}

	SDL_free(localeList);

	return prefersEnglish;
}

GameLanguageID GetBestLanguageIDFromSystemLocale(void)
{
	GameLanguageID languageID = LANGUAGE_ENGLISH;

	int numLocales = 0;
	SDL_Locale** localeList = SDL_GetPreferredLocales(&numLocales);

	for (int locale = 0; locale < numLocales; locale++)
	{
		const char* localeLanguage = localeList[locale]->language;

		for (int language = 0; language < NUM_LANGUAGES; language++)
		{
			if (0 == SDL_strncmp(localeLanguage, kLanguageCodesISO639_1[language], 2))
			{
				languageID = language;
				goto foundLocale;
			}
		}
	}

foundLocale:
	SDL_free(localeList);

	return languageID;
}

File: Source/System/Localization.swift
// Localization.swift - Port of Localization.c to Swift (partial)
//
// LocalizeWithPlaceholder stays in Localization.c: it takes C variadic
// arguments (...), which Swift can't implement. IsNativeEnglishSystem and
// GetBestLanguageIDFromSystemLocale also stay in C since they consume
// kLanguageCodesISO639_1, a 2D fixed char array that isn't practical to
// build from Swift.

private let kCsvPath = ":System:strings.csv"

private let kNullPlaceholder = strdup("???")
private let kNotLoadedMsg = strdup("STRINGS NOT LOADED")
private let kIllegalIdMsg = strdup("ILLEGAL STRING ID")
private let kEmptyMsg = strdup("")

/// Localized-string table (backs the C-ABI Localize). Owned by GameEngine
/// as `gEngine.localization`.
final class LocalizationSystem {
    fileprivate var currentLanguage: GameLanguageID = LANGUAGE_ILLEGAL
    fileprivate var stringsBuffer: UnsafeMutablePointer<CChar>?
    fileprivate var stringsTable: [UnsafeMutablePointer<CChar>?] = []
}

func LoadLocalizedStrings(_ languageID: GameLanguageID) {
    // Don't bother reloading strings if we've already loaded this language
    if languageID == gEngine.localization.currentLanguage {
        return
    }

    // Free previous strings buffer
    DisposeLocalizedStrings()

    SwGameAssert(languageID.rawValue >= 0)
    SwGameAssert(languageID.rawValue < NUM_LANGUAGES.rawValue)

    var count: Int = 0
    gEngine.localization.stringsBuffer = LoadTextFile(kCsvPath, &count)

    let maxStrings = Int(NUM_LOCALIZED_STRINGS.rawValue) + 1
    gEngine.localization.stringsTable = [UnsafeMutablePointer<CChar>?](repeating: nil, count: maxStrings)
    gEngine.localization.stringsTable[Int(STR_NULL.rawValue)] = kNullPlaceholder
    SwGameAssert(STR_NULL.rawValue == 0) // STR_NULL must be 0!

    var row: Int32 = 1 // start row at 1, so that 0 is an illegal index (STR_NULL)

    var csvReader: UnsafeMutablePointer<CChar>? = gEngine.localization.stringsBuffer
    while csvReader != nil {
        var myPhrase: UnsafeMutablePointer<CChar>?
        var eol = false

        var x: Int32 = 0
        while !eol {
            let phrase = CSVIterator(&csvReader, &eol)

            if let phrase, phrase.pointee != 0, (x == languageID.rawValue || myPhrase == nil) {
                myPhrase = phrase
            }
            x += 1
        }

        if let myPhrase {
            SwGameAssert(row < NUM_LOCALIZED_STRINGS.rawValue)
            gEngine.localization.stringsTable[Int(row)] = myPhrase
            row += 1
        }
    }

    SwGameAssert(row == NUM_LOCALIZED_STRINGS.rawValue)
}

/// Swift-native accessor over the same string table the C-ABI Localize
/// serves (LocalizeWithPlaceholder in Localization.c is the one remaining
/// C caller that needs the raw pointer form).
func localized(_ stringID: LocStrID) -> String {
    String(cString: Localize(stringID))
}

@c @implementation
public func Localize(_ stringID: LocStrID) -> UnsafePointer<CChar> {
    guard gEngine.localization.stringsBuffer != nil else { return UnsafePointer(kNotLoadedMsg!) }

    let id = Int(stringID.rawValue)
    guard id >= 0 && id < gEngine.localization.stringsTable.count else { return UnsafePointer(kIllegalIdMsg!) }

    guard let entry = gEngine.localization.stringsTable[id] else { return UnsafePointer(kEmptyMsg!) }

    return UnsafePointer(entry)
}

func DisposeLocalizedStrings() {
    if gEngine.localization.stringsBuffer != nil {
        SafeDisposePtr(gEngine.localization.stringsBuffer)
        gEngine.localization.stringsBuffer = nil
    }
}
